pkg/boot/syslinux/syslinux.go

```diff
@@ -219,29 +317,95 @@ func (c *parser) append(ctx context.Context, config string) error {
 			c.labelOrder = append(c.labelOrder, c.curEntry)
 
 		case "kernel":
-			k, err := c.getFile(arg)
-			if err != nil {
-				return err
+			// I hate special cases like these, but we aren't gonna
+			// implement syslinux modules.
+			if arg == "mboot.c32" {
+				// Prepare for a multiboot kernel.
+				delete(c.linuxEntries, c.curEntry)
+				c.mbEntries[c.curEntry] = &boot.MultibootImage{
+					Name: c.curEntry,
+				}
+			}
+			fallthrough
+
+		case "linux":
+			if e, ok := c.linuxEntries[c.curEntry]; ok {
+				k, err := c.getFile(arg)
+				if err != nil {
+					return err
+				}
+				e.Kernel = k
 			}
-			c.linuxEntries[c.curEntry].Kernel = k
 
 		case "initrd":
-			i, err := c.getFile(arg)
-			if err != nil {
-				return err
+			if e, ok := c.linuxEntries[c.curEntry]; ok {
+				// TODO: support multiple comma-separated initrds.
+				// TODO: append "initrd=$arg" to the cmdline.
+				//
+				// For how this interacts with global appends,
+				// read
+				// https://wiki.syslinux.org/wiki/index.php?title=Directives/append
+				i, err := c.getFile(arg)
+				if err != nil {
+					return err
+				}
+				e.Initrd = i
 			}
-			c.linuxEntries[c.curEntry].Initrd = i
 
 		case "append":
 			switch c.scope {
 			case scopeGlobal:
 				c.globalAppend = arg
 
 			case scopeEntry:
-				if arg == "-" {
-					c.linuxEntries[c.curEntry].Cmdline = ""
-				} else {
-					c.linuxEntries[c.curEntry].Cmdline = arg
+				if e, ok := c.mbEntries[c.curEntry]; ok {
+					modules := strings.Split(arg, "---")
+					// The first module is special -- the kernel.
+					if len(modules) > 0 {
+						kernel := strings.Fields(modules[0])
+						k, err := c.getFile(kernel[0])
+						if err != nil {
+							return err
+						}
+						e.Kernel = k
+						if len(kernel) > 0 {
+							e.Cmdline = strings.Join(kernel[1:], " ")
+						}
+						modules = modules[1:]
+					}
+					for _, cmdline := range modules {
+						m := strings.Fields(cmdline)
+						if len(m) == 0 {
+							continue
+						}
+						name := m[0]
+						file, err := c.getFile(name)
+						if err != nil {
+							return err
+						}
+						e.Modules = append(e.Modules, multiboot.Module{
+							CmdLine: strings.Join(m, " "),
```

Comment:
Is including the module name in cmdline intentional?

Reply:
actually, isn't this just cmdline? why do you need to join them back? are you worried about other forms of separators like commas

Reply:
you've got a point -- i did add trimming of spaces before and after because we split on `---` and in stuff like `xen.gz --- vmlinuz foobar --- initrd.gz` that makes cmdline ` vmlinuz foobar ` with the space before and after.

Reply:
also, yeah, the name belongs in the cmdline. it's like an execve in Linux -- arg[0] is the name. (we previously did this wrong in ESXi and chopped them off, actually, see #1675)